### PR TITLE
Add flag to enable/disable host endpoint policy on forward traffic.

### PR DESCRIPTION
**Pull request by song-jiang**
## Description
This feature allow user to specify if a host endpoint policy applies on forward traffic or not. Currently host endpoint policy applies on forward traffic regardlessly. 

## Todos
- [x] Unit tests (full coverage)
- [x] Integration tests
- [ ] Documentation
- [ ] Release note

### CODE REVIEW

File: calc/policy_sorter.go [hunk 1]
```diff
@@ -112,6 +113,20 @@ func (p PolKV) String() string {
 	return fmt.Sprintf("%s(%s)", p.Key.Name, orderStr)
 }
 
+func (p *PolKV) GovernsApplyOnForward() bool {
```
Comment (fasaxc) [inline calc/policy_sorter.go]: Naming: `GovernsApplyOnForward` doesn't read so well.  In this context, "Governs" means "Has control over".  "Has control over ingress (traffic)" makes sense but "Has control over apply on forward" doesn't.  

How about `AppliesOnForward` or `GovernsForwardedTraffic`?

File: calc/policy_sorter.go [hunk 2]
```diff
@@ -61,6 +61,7 @@ func (poc *PolicySorter) OnUpdate(update api.Update) (dirty bool) {
 				oldPolicy.Order != newPolicy.Order ||
 				oldPolicy.DoNotTrack != newPolicy.DoNotTrack ||
 				oldPolicy.PreDNAT != newPolicy.PreDNAT ||
+				*oldPolicy.ApplyOnForward != *newPolicy.ApplyOnForward ||
```
Comment (fasaxc) [inline calc/policy_sorter.go]: I think this will panic if one of the fields is nil

File: rules/endpoints.go [hunk 3]
```diff
@@ -89,6 +91,30 @@ func (r *DefaultRuleRenderer) HostEndpointToFilterChains(
...
+			ingressForwardPolicyNames,
+			profileIDs,
+			ifaceName,
+			PolicyInboundPfx,
+			ProfileInboundPfx,
+			HostFromEndpointForwardPfx,
+			ChainFailsafeIn,
+			chainTypeNormal,
+			true, // Host endpoints are always admin up.
+		),
```
Comment (fasaxc) [inline rules/endpoints.go]: TODO: handle the change from default deny to default allow?

File: rules/static.go [hunk 4]
```diff
@@ -289,14 +289,13 @@ func (r *DefaultRuleRenderer) StaticFilterForwardChains() []*Chain {
 			Action: ClearMarkAction{Mark: r.allCalicoMarkBits() &^ r.IptablesMarkAccept},
 		},
 		Rule{
-			// Unless the packet has already been accepted by untracked or pre-DNAT
-			// processing, apply normal policy for the incoming host endpoint.
+			// Apply forward policy for the incoming Host endpoint.
 			Match:  Match().MarkClear(r.IptablesMarkAccept),
-			Action: JumpAction{Target: ChainDispatchFromHostEndpoint},
+			Action: JumpAction{Target: ChainDispatchFromHostEndPointForward},
```
Comment (fasaxc) [inline rules/static.go]: TODO: apply forward policy before workload endpoint policy so forward policy applies to local workload traffic too?

File: calc/policy_sorter.go [hunk 5]
```diff
@@ -112,6 +113,20 @@ func (p PolKV) String() string {
...
 
+func (p *PolKV) GovernsApplyOnForward() bool {
+	if p.Value != nil {
+		if p.Value.ApplyOnForward != nil {
+			return *p.Value.ApplyOnForward
+		}
+
+		// If ApplyOnForward flag pointer is nil, this policy has been created by old version of calico.
+		// Apply on forward traffic as default.
+		return true
```
Comment (fasaxc) [inline calc/policy_sorter.go]: TODO: Think we've decided to change this default.

File: calc/policy_sorter.go [hunk 6]
```diff
@@ -112,6 +113,20 @@ func (p PolKV) String() string {
 	return fmt.Sprintf("%s(%s)", p.Key.Name, orderStr)
 }
 
+func (p *PolKV) GovernsApplyOnForward() bool {
+	if p.Value != nil {
```
Comment (fasaxc) [inline calc/policy_sorter.go]: Minor: think the golang convention is to handle the nils first with an early return:
```#!go
if p.Value == nil {
    return false
}
if p.Value.ApplyOnForward == nil {
    // If ApplyOnForward flag pointer is nil, this policy has been created by old version of calico.
    // Apply on forward traffic as default.
    return true
}
return *p.Value.ApplyOnForward
```

**Comment by song-jiang** [comment]
changes from last commit. 1. use boolean instead of boolean pointer. 2. rearrange kernel chain for local forward traffic. 3. Added flag to hostDispatchChains to indicate preDNAT/Untracked/ApplyOnForward chains. 4. Default allow for forward chain. 5. Completed UT.

File: rules/static.go [hunk 7]
```diff
@@ -244,7 +244,20 @@ func (r *DefaultRuleRenderer) StaticFilterForwardChains() []*Chain {
 	// The second batch actually accepts the packets if they passed through the workload policy
```
Comment (fasaxc) [inline rules/static.go]: This comment is now in the wrong place.

File: rules/rule_defs.go [hunk 8]
```diff
@@ -142,12 +146,14 @@ type RuleRenderer interface {
 		profileIDs []string,
 	) []*iptables.Chain
 
-	HostDispatchChains(map[string]proto.HostEndpointID) []*iptables.Chain
+	HostDispatchChains(map[string]proto.HostEndpointID, bool) []*iptables.Chain
```
[the remark this answers is not in the document]
Comment (song-jiang) [reply rules/rule_defs.go]: Leave it is for the moment.

File: rules/rule_defs.go [hunk 9]
```diff
@@ -71,14 +71,18 @@ const (
...
+	ChainDispatchFromHostEndPointForward = ChainNamePrefix + "from-hep-forward"
 
 	WorkloadToEndpointPfx   = ChainNamePrefix + "tw-"
 	WorkloadFromEndpointPfx = ChainNamePrefix + "fw-"
 
-	HostToEndpointPfx   = ChainNamePrefix + "th-"
-	HostFromEndpointPfx = ChainNamePrefix + "fh-"
+	HostToEndpointPfx          = ChainNamePrefix + "th-"
+	HostFromEndpointPfx        = ChainNamePrefix + "fh-"
+	HostToEndpointForwardPfx   = ChainNamePrefix + "thfw-"
```
[the remark this answers is not in the document]
Comment (song-jiang) [reply rules/rule_defs.go]: I see. We are close but should be fine.

**Pull request merged**